File: binary_search/bsearch.py
def bsearch(s, e, first, last, calls):
    print first, last, calls
    if (last - first) < 2: return s[first] == e or s[last] == e
    mid = first + (last - first)/2
    if s[mid] == e: return True
    if s[mid] > e: return bsearch(s, e, first, mid - 1, calls+1)
    else:return bsearch(s, e, mid + 1, last, calls + 1)

def search(s, e):
    print bsearch(s, e, 0, len(s) - 1, 1)

def sort(a):
    for j in range(1,len(a)):
        key = a[j]
        i = j-1
        while i>=0 and a[i]>key:
            a[i+1]=a[i]
            i=i-1
        a[i+1]=key
    return a

p = [5,4,1,3,2,6,8,13,25,78,65]
t = sort(p)

search(t,78)

#Element must give in sorted oder for using binary search

File: binary_search/bsearch.clj
(ns binary-search.core
  (:gen-class))

(defn bsearch1
  [k list]
  (if (= (count list) 1)
    (do (if (= (first list) k)
          (println "Element" k "is present in list")
          (println "Element" k "isn't present in list")))
    (let [mid (/ (count list) 2)
          current (nth list mid)
          split-list (split-at mid list)
          left (first split-list)
          right (last split-list)]
      (cond (= k current) (println "Element" k "is present in list")
            (< k current) (bsearch1 k left)
            :else (bsearch1 k right)))))

(defn bsearch2
  [k list]
  (loop [new-list list]
    (if (= (count new-list) 1)
      (do (if (= (first new-list) k)
            (println "Element" k "is present in list")
            (println "Element" k "isn't present in list")))
      (let [size (count new-list)
            mid (/ size 2)
            current (nth new-list mid)
            split-list (split-at mid new-list)
            left (first split-list)
            right (last split-list)]
        (cond (= k current) (println "Element" k "is present in list")
              (< k current) (recur left)
              :else (recur right))))))

(bsearch1 4 '(1 2 3 4 5 6 7 8 9 10))
(bsearch2 4 '(1 2 3 4 5 6 7 8 9 10))
